0.13.0-beta2 crash report

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

``` 
Terraform v0.13.0-beta2
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```
hcl https://1drv.ms/u/s!AjA715bdFDs48CF142ZuUVD1IpOW?e=7nuVZT
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->
Placed two crash logs with two different command line inputs to test if crash is repeatable.  It was.

### Expected Behavior
<!--
What should have happened?
-->

### Actual Behavior
<!--
What actually happened?
-->

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
terraform init
### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->
I am just learning Terraform and so may have invalid syntax.  

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

crash.log:

```
2020/06/24 16:57:43 [INFO] Terraform version: 0.13.0 beta2 
2020/06/24 16:57:43 [INFO] Go runtime version: go1.14.2
2020/06/24 16:57:43 [INFO] CLI args: []string{"C:\\Bin\\terraform.exe", "init", "--upgrade"}
2020/06/24 16:57:43 [DEBUG] Attempting to open CLI config file: C:\Users\Chris\AppData\Roaming\terraform.rc
2020/06/24 16:57:43 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2020/06/24 16:57:43 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2020/06/24 16:57:43 [DEBUG] ignoring non-existing provider search directory C:\Users\Chris\AppData\Roaming\terraform.d\plugins
2020/06/24 16:57:43 [DEBUG] ignoring non-existing provider search directory C:\Users\Chris\AppData\Roaming\HashiCorp\Terraform\plugins
2020/06/24 16:57:43 [INFO] CLI command args: []string{"init", "--upgrade"}
panic: not a string

goroutine 1 [running]:
github.com/zclconf/go-cty/cty.Value.AsString(0x2d28f40, 0x4167678, 0x231ba40, 0x4167200, 0x27131ae, 0x6)
	/home/circleci/project/project/vendor/github.com/zclconf/go-cty/cty/value_ops.go:1181 +0x1aa
github.com/hashicorp/terraform/configs.decodeRequiredProvidersBlock(0xc000360000, 0x272c6da, 0x12, 0x1, 0x0)
	/home/circleci/project/project/configs/provider_requirements.go:75 +0xe2a
github.com/hashicorp/terraform/configs.(*Parser).loadConfigFile(0xc00035e0a0, 0xc0004e8788, 0x7, 0x0, 0x1, 0x0, 0x0, 0x4)
	/home/circleci/project/project/configs/parser_config.go:76 +0x1b83
github.com/hashicorp/terraform/configs.(*Parser).LoadConfigFile(...)
	/home/circleci/project/project/configs/parser_config.go:22
github.com/hashicorp/terraform/configs.(*Parser).loadFiles(0xc00035e0a0, 0xc00035e140, 0x2, 0x2, 0x0, 0x2, 0x0, 0x0, 0x0, 0x0, ...)
	/home/circleci/project/project/configs/parser_config_dir.go:77 +0x349
github.com/hashicorp/terraform/configs.(*Parser).LoadConfigDir(0xc00035e0a0, 0x270e485, 0x1, 0x0, 0x1, 0x2, 0x0)
	/home/circleci/project/project/configs/parser_config_dir.go:37 +0x128
github.com/hashicorp/terraform/command.(*Meta).loadSingleModule(0xc0000a8000, 0x270e485, 0x1, 0x0, 0x0, 0x0, 0xe)
	/home/circleci/project/project/command/meta_config.go:113 +0x13c
github.com/hashicorp/terraform/command.(*InitCommand).Run(0xc0000a8000, 0xc0000560a0, 0x1, 0x2, 0xc0002cb560)
	/home/circleci/project/project/command/init.go:151 +0xa59
github.com/mitchellh/cli.(*CLI).Run(0xc000453b80, 0xc000453b80, 0xc00063fd48, 0x1)
	/home/circleci/project/project/vendor/github.com/mitchellh/cli/cli.go:255 +0x1e1
main.wrappedMain(0x0)
	/home/circleci/project/project/main.go:275 +0xc97
main.realMain(0x0)
	/home/circleci/project/project/main.go:106 +0xe5
main.main()
	/home/circleci/project/project/main.go:42 +0x41
```

main.tf:

```hcl
/* Bootstrap terraform with creating the AWS S3 resources needed for tracking/maintaining state.
 */
terraform {
  required_version = '~> 0.13.0'
  required_providers {
    aws = {
      source  = 'hashicorp/aws'
      version = '~> 2.67'
    }
  }
}

provider 'aws' {
  region = var.region
}

resource 'aws_s3_bucket_public_access_block' 'terraform_state' {
  bucket              = aws_s3_bucket.terraform_state.id
  block_public_acls   = true
  block_public_policy = true
  tags = {
    Name = 'table-locks'
    duty = 'devops'
  }
}

resource 'aws_s3_bucket_public_access_block' 'log_sink' {
  bucket              = aws_s3_bucket.log-sink.id
  block_public_acls   = true
  block_public_policy = true
}

resource 'aws_s3_bucket' 'terraform_state' {
  bucket = 'tycarac-devops'
  acl    = 'private'
  lifecycle {
    prevent_destroy = true
  }
  versioning {
    enabled    = true
    mfa_delete = false
  }
  logging {
    target_bucket = aws_s3_bucket.log-sink.id
    target_prefix = 'log/'
  }
  server_side_encryption_configuration {
    rule {
      apply_server_side_encryption_by_default {
        sse_algorithm = 'AES256'
      }
    }
  }
}

resource 'aws_s3_bucket' 'log-sink' {
  bucket = 'tycarac-logging'
  acl    = 'log-delivery-write'
  lifecycle {
    prevent_destroy = true
  }
  lifecycle_rule {
    id      = 'log'
    enabled = true
    prefix  = 'log/'
    expiration {
      days = 65
    }
    tags = {
      product   = 'devops'
      autoclean = true
    }
  }
  server_side_encryption_configuration {
    rule {
      apply_server_side_encryption_by_default {
        sse_algorithm = 'AES256'
      }
    }
  }
}

resource 'aws_dynamodb_table' 'tycarac-devops-locks' {
   name = 'terraform-lock'
   hash_key = 'LockID'
   read_capacity = 5
   write_capacity = 5

   attribute {
      name = 'LockID'
      type = 'S'
   }

   tags = {
     Name = 'Terraform Lock Table'
   }
}
```

--- a/configs/provider_requirements.go
+++ b/configs/provider_requirements.go
@@ -4,6 +4,7 @@ import (
 	version "github.com/hashicorp/go-version"
 	"github.com/hashicorp/hcl/v2"
 	"github.com/hashicorp/terraform/addrs"
+	"github.com/zclconf/go-cty/cty"
 )
 
 // RequiredProvider represents a declaration of a dependency on a particular
@@ -55,41 +56,61 @@ func decodeRequiredProvidersBlock(block *hcl.Block) (*RequiredProviders, hcl.Dia
 				vc := VersionConstraint{
 					DeclRange: attr.Range,
 				}
-				constraintStr := expr.GetAttr("version").AsString()
-				constraints, err := version.NewConstraint(constraintStr)
-				if err != nil {
-					// NewConstraint doesn't return user-friendly errors, so we'll just
-					// ignore the provided error and produce our own generic one.
+				constraint := expr.GetAttr("version")
+				if !constraint.Type().Equals(cty.String) || constraint.IsNull() {
 					diags = append(diags, &hcl.Diagnostic{
 						Severity: hcl.DiagError,
 						Summary:  "Invalid version constraint",
-						Detail:   "This string does not use correct version constraint syntax.",
+						Detail:   "Version must be specified as a string.",
 						Subject:  attr.Expr.Range().Ptr(),
 					})
 				} else {
-					vc.Required = constraints
-					rp.Requirement = vc
+					constraintStr := constraint.AsString()
+					constraints, err := version.NewConstraint(constraintStr)
+					if err != nil {
+						// NewConstraint doesn't return user-friendly errors, so we'll just
+						// ignore the provided error and produce our own generic one.
+						diags = append(diags, &hcl.Diagnostic{
+							Severity: hcl.DiagError,
+							Summary:  "Invalid version constraint",
+							Detail:   "This string does not use correct version constraint syntax.",
+							Subject:  attr.Expr.Range().Ptr(),
+						})
+					} else {
+						vc.Required = constraints
+						rp.Requirement = vc
+					}
 				}
 			}
 			if expr.Type().HasAttribute("source") {
-				rp.Source = expr.GetAttr("source").AsString()
-
-				fqn, sourceDiags := addrs.ParseProviderSourceString(rp.Source)
-
-				if sourceDiags.HasErrors() {
-					hclDiags := sourceDiags.ToHCL()
-					// The diagnostics from ParseProviderSourceString don't contain
-					// source location information because it has no context to compute
-					// them from, and so we'll add those in quickly here before we
-					// return.
-					for _, diag := range hclDiags {
-						if diag.Subject == nil {
-							diag.Subject = attr.Expr.Range().Ptr()
-						}
-					}
-					diags = append(diags, hclDiags...)
+				source := expr.GetAttr("source")
+				if !source.Type().Equals(cty.String) || source.IsNull() {
+					diags = append(diags, &hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Invalid source",
+						Detail:   "Source must be specified as a string.",
+						Subject:  attr.Expr.Range().Ptr(),
+					})
 				} else {
-					rp.Type = fqn
+					rp.Source = source.AsString()
+
+					fqn, sourceDiags := addrs.ParseProviderSourceString(rp.Source)
+
+					if sourceDiags.HasErrors() {
+						hclDiags := sourceDiags.ToHCL()
+						// The diagnostics from ParseProviderSourceString don't contain
+						// source location information because it has no context to compute
+						// them from, and so we'll add those in quickly here before we
+						// return.
+						for _, diag := range hclDiags {
+							if diag.Subject == nil {
+								diag.Subject = attr.Expr.Range().Ptr()
+							}
+						}
+						diags = append(diags, hclDiags...)
+					} else {
+						rp.Type = fqn
+					}
 				}
 			}
 

--- a/configs/provider_requirements_test.go
+++ b/configs/provider_requirements_test.go
@@ -306,6 +306,32 @@ func TestDecodeRequiredProvidersBlock(t *testing.T) {
 			},
 			Error: "Invalid required_providers syntax",
 		},
+		"invalid source attribute type": {
+			Block: &hcl.Block{
+				Type: "required_providers",
+				Body: hcltest.MockBody(&hcl.BodyContent{
+					Attributes: hcl.Attributes{
+						"my-test": {
+							Name: "my-test",
+							Expr: hcltest.MockExprLiteral(cty.ObjectVal(map[string]cty.Value{
+								"source": cty.DynamicVal,
+							})),
+						},
+					},
+				}),
+				DefRange: blockRange,
+			},
+			Want: &RequiredProviders{
+				RequiredProviders: map[string]*RequiredProvider{
+					"my-test": {
+						Name:      "my-test",
+						DeclRange: mockRange,
+					},
+				},
+				DeclRange: blockRange,
+			},
+			Error: "Invalid source",
+		},
 	}
 
 	for name, test := range tests {

